src/persist-client/src/batch.rs

```diff
@@ -720,9 +742,10 @@ where
             if let Some((max_run_k, max_run_v, max_run_t)) = &mut self.max_kvt_in_run {
                 // start a new run if our part contains an update that exists in the
                 // range already covered by the existing parts of the current run
-                if (min_part_k, min_part_v, &min_part_t) < (max_run_k, max_run_v, max_run_t) {
-                    self.runs.push(self.parts_written);
-                }
+                soft_assert_or_log!(
+                    (min_part_k, min_part_v, &min_part_t) >= (max_run_k, max_run_v, max_run_t),
+                    "expected data in sorted order"
+                );
```

Comment:
It feels like this should be an `assert!` or we should retain the original if-statement and move the `soft_assert!` in there?

AFAIU `self.runs` is a list of offsets, e.g. [0, 5, 7] indicates that there are 2 Runs, the first consists of parts `0..5`, the second of parts `5..7`. Because this is a soft_assert it means we will never trigger this in production, which makes it possible to have a Run that is not internally sorted? Which seems bad?

I vaguely remember some part of Persist is self-correcting, so maybe this isn't bad!

Reply:
> AFAIU self.runs is a list of offsets, e.g. [0, 5, 7] indicates that there are 2 Runs, the first consists of parts 0..5, the second of parts 5..7.

Right! Though the 0 is usually implicit.

> Because this is a soft_assert it means we will never trigger this in production, which makes it possible to have a Run that is not internally sorted? Which seems bad?

Yeah, it's not great!

Right now the only code that depends on this for correctness is the fast-path peek optimization. It's also likely to trip asserts elsewhere if the data is unsorted - eg. during a future run of compaction.

The original thought was: if we got this wrong and get alerted for an error log, we might prefer to disable fast-path peeks in that env instead of experiencing downtime. But you could argue that both ways...

One option would be to do _both_... do the soft-assert and then split off a new run the way we did before. That way we don't change behaviour, but we'd get alerted (via staging panics or prod sentry) if things aren't behaving the way they're supposed to... WDYT?

Reply:
> One option would be to do both... do the soft-assert and then split off a new run the way we did before. That way we don't change behaviour, but we'd get alerted (via staging panics or prod sentry) if things aren't behaving the way they're supposed to... WDYT?

I like this approach! It allows us to validate things are working as intended and gives us an automatic escape hatch if there's an edge case in production we don't know about

Reply:
Done - thank you!